File: src/main/java/kuc/karol/algorithms/java/bst/TreeNode.java
package kuc.karol.algorithms.java.bst;

import java.util.Objects;
import java.util.stream.Stream;

public class TreeNode {
    private TreeNode left;
    private TreeNode right;
    private Integer value;

    public TreeNode(TreeNode left, TreeNode right, Integer value) {
        this.left = left;
        this.right = right;
        this.value = value;
    }

    public TreeNode(Integer value) {
        this(null, null, value);
    }

    public TreeNode left() {
        return left;
    }

    public TreeNode right() {
        return right;
    }

    public Integer value() {
        return value;
    }

    public TreeNode subtreeFor(Target target) {
        if (value.equals(target.value())) {
            return this;
        } else if (target.value() < value) {
            return left;
        } else {
            return right;
        }
    }

    public Integer findClosest(Target target) {
        Closest closest = new Closest(value);
        TreeNode node = this;

        while (target.canHaveBetterMatchIn(node, closest)) {
            Closest current = new Closest(node.value());
            if (current.isCloserToTargetThan(closest, target)) {
                closest = current;
            }
            node = node.subtreeFor(target);
        }
        return closest.value();
    }

    void insertRecursive(Integer value) {
        if (value < this.value) {
            if (this.left == null) {
                this.left = new TreeNode(value);
            } else {
                this.left.insertRecursive(value);
            }
        } else if (value > this.value) {
            if (this.right == null) {
                this.right = new TreeNode(value);
            } else {
                this.right.insertRecursive(value);
            }
        }
    }

    TreeNode delete(Integer toDelete) {
        if (toDelete < value && left != null) {
            left = left.delete(toDelete);
            return this;
        }
        if (toDelete > value && right != null) {
            right = right.delete(toDelete);
            return this;
        }
        // If we reach here and the value doesn't match, the node to delete doesn't exist
        if (toDelete != value) {
            return this; // Return unchanged if node to delete is not found
        }
        if (hasNoChildren()) return null;
        var children = Stream.of(left, right).filter(Objects::nonNull).toList();
        if (children.size() == 1) {
            return children.getFirst();
        }
        TreeNode successor = null;
        if (right() != null) {
            successor = right.findNodeWithSmallestValue();
            if(successor == right) {
                this.right = null;
            } else {
                this.right = this.right.delete(successor.value);
            }
            this.value = successor.value;
        }

        return this;
    }

    private boolean hasNoChildren() {
        return left == null && right == null;
    }


    TreeNode findNodeWithSmallestValue() {
        TreeNode current = this;
        while (current.left != null) {
            current = current.left;
        }
        return current;
    }

    TreeNode findNodeWithBiggestValue() {
        TreeNode current = this;
        while (current.right != null) {
            current = current.right;
        }
        return current;
    }
}

File: src/main/kotlin/kuc/karol/algorithms/kotlin/bst/TreeNode.kt
package kuc.karol.algorithms.kotlin.bst

import kotlin.math.abs

data class Target(val value: Int)

data class Closest(val value: Int) {
    infix fun canBeImprovedTowards(target: Target): ClosestTargetPair {
        return ClosestTargetPair(this, target)
    }

    fun isCloserThan(closest: Closest, target: Target) =
        abs(value - target.value) < abs(closest.value - target.value)

    infix fun isCloserTo(target: Target): ClosestTargetComparison {
        return ClosestTargetComparison(this, target)
    }
}

data class ClosestTargetPair(val closest: Closest, val target: Target) {
    infix fun within(node: TreeNode?): Boolean {
        return node != null && target.value != node.value
    }
}

data class ClosestTargetComparison(val closest: Closest, val target: Target) {
    infix fun than(otherClosest: Closest): Boolean {
        return closest.isCloserThan(otherClosest, target)
    }
}

data class TreeNode(
    val left: TreeNode? = null,
    val right: TreeNode? = null,
    val value: Int
) {
    fun insertRecursive(newValue: Int): TreeNode = when {
        newValue < value -> copy(left = left?.insertRecursive(newValue) ?: TreeNode(value = newValue))
        newValue > value -> copy(right = right?.insertRecursive(newValue) ?: TreeNode(value = newValue))
        else -> this
    }

    fun findClosestIterative(node: TreeNode, target: Target): Int {
        var currentNode: TreeNode? = node
        var closest = Closest(node.value)
        while (closest canBeImprovedTowards target within currentNode) {
            val maybeCloser = Closest(currentNode!!.value)
            if (maybeCloser isCloserTo target than closest) {
                closest = maybeCloser
            }
            currentNode = currentNode traverseDownTowards target
        }
        return closest.value
    }

    private infix fun traverseDownTowards(target: Target) =
        if (target.value == value) this
        else if (target.value < value) left
        else right


    companion object {
        fun insertRecursive(root: TreeNode?, value: Int): TreeNode {
            return root?.insertRecursive(value) ?: TreeNode(value = value)
        }
    }
}
